crates/circuit/src/circuit_data.rs

```diff
@@ -504,6 +504,25 @@ impl CircuitData {
         Ok(())
     }
 
+    /// Lists all operations along with the qubits involved in those operations.
+    ///
+    /// Returns:
+    ///     list.
+    #[pyo3(signature = ())]
+    pub fn operations_with_qubits(&self, py: Python<'_>) -> PyResult<Py<PyAny>> {
+        let mut ret: Vec<(Option<String>, Option<String>)> = vec![];
+        for (_index, inst) in self.data.iter().enumerate() {
+            let op_str = inst.op.view().name().to_string();
```

Comment:
We should probably try to avoid `to_string()` here as this will make a copy of every string, and we can keep it all by reference in Rust space. When we return to Python that will do the copying for us.